### ISSUE TITLE
Support Spring 6 HTTP interface

### ISSUE THREAD

**Issue by minwoox** [line/armeria#4838]
The Spring Framework 6 supports define declarative HTTP services using Java interfaces like [Feign](https://www.baeldung.com/spring-boot-feignclient-vs-webclient).
https://www.baeldung.com/spring-6-http-interface
~~We can also support this by providing [HttpClientAdapter](https://github.com/spring-projects/spring-framework/blob/main/spring-web/src/main/java/org/springframework/web/service/invoker/HttpClientAdapter.java)~~
It already work if WebFlux integration module is used.
```java
interface BooksService {

    @GetExchange("/books")
    List<Book> getBooks();

    @GetExchange("/books/{id}")
    Book getBook(@PathVariable long id);

    @PostExchange("/books")
    Book saveBook(@RequestBody Book book);

    @DeleteExchange("/books/{id}")
    ResponseEntity<Void> deleteBook(@PathVariable long id);
}

WebClient webClient = ... // WebFlux webClient which is created via auto configuration
HttpServiceProxyFactory httpServiceProxyFactory = HttpServiceProxyFactory
  .builder(WebClientAdaptor.forClient(webClient))
  .build();
booksService = httpServiceProxyFactory.createClient(BooksService.class);
```

**Comment by minwoox** [comment]
I realized that this already works if Armeria WebFlux integration module is used.
Please let us know if you want to use this feature without WebFlux.

**Comment by ikhoon** [comment]
We've changed our integration for Spring 6 HTTP interface to provide [HttpClientAdapter](https://github.com/spring-projects/spring-framework/blob/main/spring-web/src/main/java/org/springframework/web/service/invoker/HttpClientAdapter.java?rgh-link-date=2023-04-21T04%3A04%3A27Z). #4912

**Comment by heowc** [comment]
I‘m interested in this 🤔

**Comment by ikhoon** [comment]
Usually, we don't tell contributors a schedule, but this issue needs a schedule. We are planning to implement it in 3Q because there is a desire for this feature at LINE internally.

The non-draft PR should come out by mid-August because review time is needed to match the release schedule to 3Q. It seems possible because we have enough time. 

If you can meet this schedule, I'd like to ask you to implement this function.
Once a non-draft is out, I'll review it and make some commits if necessary.

Anyway, thanks for your interest. 🙏

**Comment by heowc** [comment]
Oh yeah. :smiley: 

I was implementing some of the features, but I think they could be better implemented if we reviewed them internally. (If he/she post a PR, I'll review it together).
- https://github.com/heowc/armeria/commit/fd1e928527b580223db45cf9fd40a3bdcc2a293e

**Comment by ikhoon** [comment]
If you already have implemented the feature. Please continue to implement and finish.

**Comment by heowc** [comment]
> We are going to add the Spring 6 HTTP interface to `armeria-spring6` module since it is not related to Spring Boot. Hence, we don't need to consider Spring Boot 3.2 schedule.

I agree 😎

**Comment by heowc** [comment]
I have a question.

For spring related clients, there are mostly internal converter services like armeria's annotation service that look at the content-type or body type and make the appropriate conversion, but for armeria clients, the user selects and converts. 

If we implement that feature, will we do it like for spring related clients? Or like armeria client?

**Comment by heowc** [comment]
> I have a question.
> 
> For spring related clients, there are mostly internal converter services like armeria's annotation service that look at the content-type or body type and make the appropriate conversion, but for armeria clients, the user selects and converts.
> 
> If we implement that feature, will we do it like for spring related clients? Or like armeria client?

This question may be related to this issue.
- https://github.com/line/armeria/issues/4910

**Comment by ikhoon** [comment]
> For spring related clients, there are mostly internal converter services like armeria's annotation service that look at the content-type or body type and make the appropriate conversion, but for armeria clients, the user selects and converts.

What `content-type`s are supported by Spring 6? As JSON is a dominant serialization format for REST API, why don't we start with JSON first? We may add an extension API to add a custom converter for requests and responses like Retrofit later.
https://github.com/line/armeria/blob/41d097efd92446d60f034cce3bee9a26f0ba47f9/retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/ArmeriaRetrofitBuilder.java#L153-L156

**Comment by jrhee17** [comment]
Linking to https://github.com/line/armeria/pull/5370

**Comment by minwoox** [comment]
Closed by #5370
